In [1]:
import numpy as np
import torch
import torch.nn as nn # used for training and building neural networks
import torch.optim as optim # for otpimizers

# Introduction

## Training Loops

In [6]:
# Data
xs = torch.tensor(np.arange(50).reshape(50, 1), dtype=torch.float32)
ys = 2 * xs + 1

# model
# sequential model, input goes thru layers sequentially
# model has only one layer, where it accepts an input of size 1
# and outputs a size 1 tensor
model = nn.Sequential(nn.Linear(1, 1))
# Linear layer so its a y = np.dot(weights.T, xs) + b

# Loss function
criterion = nn.MSELoss() # mean square error

# optimizer
# choose an optimizer then within it always put model.parameters()
optimizer = optim.SGD(model.parameters(), lr = 1e-5)
# the optimizer updates the weights 
# in the direction of the gradient vector

# for using the gpu
device = torch.device("cuda:0")
model = model.to(device)

# Training Loop
for _ in range(500):
    model.train()

    # moves dataset to gpu
    xs = xs.to(device)
    ys = ys.to(device)
    
    optimizer.zero_grad() # this zeros the gradient for every loop
    
    outputs = model(xs)
    loss = criterion(outputs, ys)
    print(f"Epoch {_}: {loss}")
    
    loss.backward() 
    # this finds the gradient vector, 
    # partial derivative of loss_fn with respect to weights
    
    optimizer.step()

Epoch 0: 6456.7099609375
Epoch 1: 6249.400390625
Epoch 2: 6048.748046875
Epoch 3: 5854.5380859375
Epoch 4: 5666.5634765625
Epoch 5: 5484.6240234375
Epoch 6: 5308.52734375
Epoch 7: 5138.08349609375
Epoch 8: 4973.11279296875
Epoch 9: 4813.43896484375
Epoch 10: 4658.8916015625
Epoch 11: 4509.306640625
Epoch 12: 4364.5244140625
Epoch 13: 4224.390625
Epoch 14: 4088.756103515625
Epoch 15: 3957.477294921875
Epoch 16: 3830.4130859375
Epoch 17: 3707.4287109375
Epoch 18: 3588.392822265625
Epoch 19: 3473.178955078125
Epoch 20: 3361.66455078125
Epoch 21: 3253.73046875
Epoch 22: 3149.26220703125
Epoch 23: 3048.14794921875
Epoch 24: 2950.2802734375
Epoch 25: 2855.554931640625
Epoch 26: 2763.870849609375
Epoch 27: 2675.130615234375
Epoch 28: 2589.23974609375
Epoch 29: 2506.1064453125
Epoch 30: 2425.642578125
Epoch 31: 2347.76220703125
Epoch 32: 2272.382080078125
Epoch 33: 2199.422607421875
Epoch 34: 2128.8056640625
Epoch 35: 2060.4560546875
Epoch 36: 1994.300537109375
Epoch 37: 1930.269775390625
Epoc

## Evaluation

In [7]:
model.eval()
x_k = torch.tensor([[-1], [100]], dtype = torch.float32).to(device)
with torch.no_grad():
    y_hat = model(x_k)
    print(y_hat)

tensor([[ -1.4372],
        [201.7826]], device='cuda:0')


## Accessing bias and weights

In [10]:
layer = model[0]
weights = layer.weight.data.cpu().numpy()
bias = layer.bias.data.cpu().numpy()

In [11]:
weights, bias

(array([[2.0120773]], dtype=float32), array([0.5748689], dtype=float32))

this means that the model learned that its y = 2.0120773 * x  + 0.5748689